### PR TITLE
Bibtex parser

### PR DESCRIPTION
**Pull request by TwlyY29**
Hi,

as a user of geany I was missing an overview of bibliographic entries when opening bibtex files for a long time. 
After contacting the maintainers of geany and proposing some code that adds the missing functionality, they recommended me to add such a parser to the ctags project as well (https://www.mail-archive.com/devel@lists.geany.org/msg03221.html). 
Well, and here I am, proposing my bibtex-parser to you guys as well. It differs from the geany implementation in that it is now based on the existing tex-parser that ships with ctags.
The result of the parser is a list of identifiers for bibtex entries.

Let me know what you think.

Best
Mirco

### CODE REVIEW

File: parsers/bibtex.c [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+	TOKEN_IDENTIFIER,
+	TOKEN_STRING,
+};
+typedef int tokenType;
+
+typedef struct sTokenInfo {
+	tokenType		type;
+	keywordId		keyword;
+	vString *		string;
+	vString *		scope;
```
Comment (masatake) [inline parsers/bibtex.c]: It seems that scope field is not used.

File: parsers/bibtex.c [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+/*
+ *	 Tag generation functions
+ */
+static void makeBibTag (tokenInfo *const token, bibKind kind)
+{
+	if (BibKinds [kind].enabled)
+	{
+		const char *const name = vStringValue (token->string);
+		int parentKind = KIND_GHOST_INDEX;
+		vString *parentName = vStringNew();
```
Comment (masatake) [inline parsers/bibtex.c]: parentName is not used. parentKnd is not uesd. Both can be remoed.

File: parsers/bibtex.c [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+	tokenInfo *const name = newToken ();
+	vString *	currentid;
+	bool		useLongName = true;
+	bool        eof = false;
+
+	currentid = vStringNew ();
+	/*
+	 * Bib entries are of these formats:
+	 *   @article{identifier,
+   *   author="John Doe"}
```
Comment (masatake) [inline parsers/bibtex.c]: It seems that indentation is a bit broken.

File: parsers/bibtex.c [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+	if (isType (token, TOKEN_OPEN_CURLY))
+	{
+		int depth = 1;
+
+		if (!readToken (token))
+		{
+			eof = true;
+			goto out;
+		}
+    if (isType (token, TOKEN_IDENTIFIER)){
```
Comment (masatake) [inline parsers/bibtex.c]: It seems that indentation is a bit broken.

File: parsers/bibtex.c [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+      vStringCat (currentid, token->string);
+      vStringStripTrailing (currentid);
+			if (vStringLength (currentid) > 0)
+			{
+				vStringCopy (name->string, currentid);
+				makeBibTag (name, kind);
+			}
+    }else{ // should find an identifier for bib item at first place
```
Comment (masatake) [inline parsers/bibtex.c]: Puts whitespaces around else. or
```
}
else
}
```
is better.

**Comment by masatake** [comment]
Thank you. I have 3 requests.

1. add test cases. See http://docs.ctags.io/en/latest/units.html . Many examples are in Units/*.
2. update win32/ctags_vs2013.vcxproj and ctags_vs2013.vcxproj.filters. An example can be found in 
https://github.com/universal-ctags/ctags/pull/2130/commits/ce3627384dc31840ada2b6ac0c5a4b7aed6ec478 .
3 add your parsre to docs/news.rst as a new parser.

File: parsers/bibtex.c [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
+						token->type = TOKEN_IDENTIFIER;
+					}
+					break;
+	}
+	return true;
+}
+
+static bool readToken (tokenInfo *const token)
+{
+	return readTokenFull (token, false);
```
Comment (masatake) [inline parsers/bibtex.c]: readTokenFull is used only here.
It means the second argument of readTokenFull is never used.
So you can move the code of readTokenFull to readToken, and you can remove readTokenFull.

File: parsers/bibtex.c [hunk 7]
```diff
@@ -0,0 +1,447 @@
...
+	} while (true);
+}
+
+static void initialize (const langType language)
+{
+	Assert (ARRAY_SIZE (BibKinds) == BIBTAG_COUNT);
+	Lang_bib = language;
+}
+
+static void finalize (const langType language CTAGS_ATTR_UNUSED,
```
Comment (masatake) [inline parsers/bibtex.c]: Do you have a plan to add lines to this function?

File: parsers/bibtex.c [hunk 8]
```diff
@@ -0,0 +1,447 @@
...
+	vStringClear (token->string);
+
+getNextChar:
+
+	do
+	{
+		c = getcFromInputFile ();
+		whitespaces++;
+	}
+	while (c == '\t'  ||  c == ' ' ||  c == '\n');
```
Comment (masatake) [inline parsers/bibtex.c]: Just about the style.
Two whitespaces are put around `||`. But, only one is needed.

File: parsers/bibtex.c [hunk 9]
```diff
@@ -0,0 +1,447 @@
...
+ *	 BibTex language "reference":
+ *		 https://en.wikipedia.org/wiki/BibTeX
+ */
+
+/*
+ *	 INCLUDE FILES
+ */
+#include "general.h"	/* must always come first */
+#include <ctype.h>	/* to define isalpha () */
+#ifdef DEBUG
```
Comment (masatake) [inline parsers/bibtex.c]: Do you need this?

File: parsers/bibtex.c [hunk 10]
```diff
@@ -0,0 +1,447 @@
...
+			}
+		}
+		if (eof)
+			break;
+	} while (true);
+}
+
+static void initialize (const langType language)
+{
+	Assert (ARRAY_SIZE (BibKinds) == BIBTAG_COUNT);
```
Comment (masatake) [inline parsers/bibtex.c]: Moving this Assert to the head of BibtexParser is better.

File: Units/parser-bibtex.r/bib-simple.d/input.bib [hunk 11]
```diff
@@ -0,0 +1,27 @@
+% this is a comment that will be ignored
```
Comment (masatake) [inline Units/parser-bibtex.r/bib-simple.d/input.bib]: I would like you to extend the input for testing all kinds.

File: parsers/bibtex.c (hunk 6 above)
Comment (TwlyY29) [reply parsers/bibtex.c]: this was removed in b87ecab

File: parsers/bibtex.c (hunk 7 above)
Comment (TwlyY29) [reply parsers/bibtex.c]: no. this was removed in 8f4ee6b

File: parsers/bibtex.c (hunk 8 above)
Comment (TwlyY29) [reply parsers/bibtex.c]: changed this to one whitespace in 8f4ee6b

File: parsers/bibtex.c (hunk 9 above)
Comment (TwlyY29) [reply parsers/bibtex.c]: no. removed the DEBUG block in 8f4ee6b

File: parsers/bibtex.c (hunk 10 above)
Comment (TwlyY29) [reply parsers/bibtex.c]: so I did (1e8c3f2)

File: Units/parser-bibtex.r/bib-simple.d/input.bib (hunk 11 above)
Comment (TwlyY29) [reply Units/parser-bibtex.r/bib-simple.d/input.bib]: extended both input.bib as well as expected_tags with all bib items in 8f4ee6b

File: parsers/bibtex.c (hunk 2 above)
Comment (TwlyY29) [reply parsers/bibtex.c]: these were removed in 952f651

File: parsers/bibtex.c (hunk 1 above)
Comment (TwlyY29) [reply parsers/bibtex.c]: you're right. I removed the field in 952f651

File: parsers/bibtex.c (hunk 3 above)
Comment (TwlyY29) [reply parsers/bibtex.c]: tabs and whitespaces got confused in my editor... I tried to fix indentation in 952f651

File: parsers/bibtex.c (hunk 4 above)
Comment (TwlyY29) [reply parsers/bibtex.c]: also 952f651

File: parsers/bibtex.c (hunk 5 above)
Comment (TwlyY29) [reply parsers/bibtex.c]: I changed it to your suggestion in 952f651

**Comment by TwlyY29** [comment]
Thanks for your thorough review and helpful hints, @masatake !

**Comment by masatake** [comment]
Thank you!

I have two extra comments.

You chose "BibTex" as the name of parser. How do you think to change it to "BibTeX" parser?

Could you consider to add "abbreviation" kind defined with `@string` keyword as described in 
http://www.bibtex.org/Format/ ?

**Comment by TwlyY29** [comment]
Hi, sure, I can rename the parser. Is it just in the bibtex.c file or somewhere else?
Also, I added the @string keyword. I'll create a separate pull request for that.

Best,
Mirco

**Pull request merged**